Automattic/jetpack
# WAF: fix normalizing nested arrays

For processing rules, we "normalize" arrays to pull out the values and keys with a prefix, optionally keeping only certain fields or removing certain fields. However, with PHP's query string magic there could be nested arrays (i.e. `?foo[1][2]=123&foo[1][3]` will turn into nested arrays in `_GET/_POST/_REQUEST`), and the existing code had a `todo` to address this. However, these nested arrays would cause errors, which this PR fixes.

Fixes 1201069996155224-as-1201890219548762

#### Changes proposed in this Pull Request:

* Flatten nested array targets as part of normalization
* Nested keys will be lost, but the keys are just used for logging anyway (at least at the moment)
* Adds unit tests for flattening and overall normalization

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:






* Add one or both of the rules below to your `rules.php` file (atlhough any rule that looks at args aka query strings should trigger it).
* Without this PR, hit a URL such as `http://localhost/?language=0&lang_settings[0][1]=http://www.qtrainer.co.kr/data/geditor/editor/banner.jpg&lang_settings[0][2]=http://www.qtrainer.co.kr/data/geditor/editor/sdgsdg.jpg` although anything with nested arrays in the query string should reproduce this.
* Verify that you get an error (TODO: our error handling code should prevent this; why didn't it?). Verify that you see an error in the logs complaining about getting an array when something else was expected.
* Apply this PR and test again; the request should work and the error logs should have no errors.
* Verify that the new unit tests pass (`composer phpunit` in this package directory).

Example rules:

```
$rule = (object) array( 'id' => 944100, 'reason' => '', 'tags' => array (
	0 => 'application-multi',
	1 => 'language-java',
	2 => 'platform-multi',
	3 => 'attack-rce',
	4 => 'owasp_crs',
	5 => 'capec/1000/152/137/6',
	6 => 'pci/6.5.2',
	7 => 'paranoia-level/1',
  ) );
  if($waf->match_targets(array (
  ),array (
	'args' => 
	array (
	),
	'args_names' => 
	array (
	),
	'request_cookies' => 
	array (
	  'except' => 
	  array (
		0 => '/__utm/',
	  ),
	),
	'request_cookies_names' => 
	array (
	),
	'request_body' => 
	array (
	),
	'request_headers' => 
	array (
	),
  ),'rx','#java\\.lang\\.(?:runtime|processbuilder)#Ds',false,false)) {
  $waf->inc_var('tx.rce_score',$waf->get_var('tx.critical_anomaly_score'));
  $waf->inc_var('tx.anomaly_score_pl1',$waf->get_var('tx.critical_anomaly_score'));
  $rule->reason = 'Remote Command Execution: Suspicious Java class detected Matched Data: '.$waf->matched_var.' found within '.$waf->matched_var_name;
  return $waf->block('block',$rule->id,$rule->reason,403);
  }

  $rule = (object) array( 'id' => 944130, 'reason' => '', 'tags' => array (
	0 => 'application-multi',
	1 => 'language-java',
	2 => 'platform-multi',
	3 => 'attack-rce',
	4 => 'owasp_crs',
	5 => 'capec/1000/152/248',
	6 => 'pci/6.5.2',
	7 => 'paranoia-level/1',
  ) );
  if($waf->match_targets(array (
  ),array (
	'args' => 
	array (
	),
	'args_names' => 
	array (
	),
	'request_cookies' => 
	array (
	  'except' => 
	  array (
		0 => '/__utm/',
	  ),
	),
	'request_cookies_names' => 
	array (
	),
	'request_body' => 
	array (
	),
	'request_filename' => 
	array (
	),
	'request_headers' => 
	array (
	),
  ),'pm',array (
	0 => 'com.opensymphony.xwork2',
	1 => 'com.sun.org.apache',
	2 => 'java.io.BufferedInputStream',
	3 => 'java.io.BufferedReader',
	4 => 'java.io.ByteArrayInputStream',
	5 => 'java.io.ByteArrayOutputStream',
	6 => 'java.io.CharArrayReader',
	7 => 'java.io.DataInputStream',
	8 => 'java.io.File',
	9 => 'java.io.FileOutputStream',
	10 => 'java.io.FilePermission',
	11 => 'java.io.FileWriter',
	12 => 'java.io.FilterInputStream',
	13 => 'java.io.FilterOutputStream',
	14 => 'java.io.FilterReader',
	15 => 'java.io.InputStream',
	16 => 'java.io.InputStreamReader',
	17 => 'java.io.LineNumberReader',
	18 => 'java.io.ObjectOutputStream',
	19 => 'java.io.OutputStream',
	20 => 'java.io.PipedOutputStream',
	21 => 'java.io.PipedReader',
	22 => 'java.io.PrintStream',
	23 => 'java.io.PushbackInputStream',
	24 => 'java.io.Reader',
	25 => 'java.io.StringReader',
	26 => 'java.lang.Class',
	27 => 'java.lang.Integer',
	28 => 'java.lang.Number',
	29 => 'java.lang.Object',
	30 => 'java.lang.Process',
	31 => 'java.lang.ProcessBuilder',
	32 => 'java.lang.reflect',
	33 => 'java.lang.Runtime',
	34 => 'java.lang.String',
	35 => 'java.lang.StringBuilder',
	36 => 'java.lang.System',
	37 => 'javax.script.ScriptEngineManager',
	38 => 'org.apache.commons',
	39 => 'org.apache.struts',
	40 => 'org.apache.struts2',
	41 => 'org.omg.CORBA',
	42 => 'java.beans.XMLDecode',
  ),false,false)) {
  $waf->inc_var('tx.rce_score',$waf->get_var('tx.critical_anomaly_score'));
  $waf->inc_var('tx.anomaly_score_pl1',$waf->get_var('tx.critical_anomaly_score'));
  $rule->reason = 'Suspicious Java class detected Matched Data: '.$waf->matched_var.' found within '.$waf->matched_var_name;
  return $waf->block('block',$rule->id,$rule->reason,403);
  }
```
